src/websocket.c

```diff
@@ -208,30 +281,41 @@ ws_execute(struct http_client *c, const char *frame, size_t frame_len) {
 	if(fun_extract) {
 
 		/* Parse websocket frame into a cmd object. */
-		struct cmd *cmd = fun_extract(c, frame, frame_len);
+		struct cmd *cmd = fun_extract(c, msg->payload, msg->payload_sz);
 
 		if(cmd) {
-			/* copy client info into cmd. */
-			cmd_setup(cmd, c);
 			cmd->is_websocket = 1;
 
-			if (c->pub_sub != NULL) {
-				/* This client already has its own connection
-				 * to Redis due to a subscription; use it from
-				 * now on. */
-				cmd->ac = c->pub_sub->ac;
-			} else if (cmd_is_subscribe(cmd)) {
-				/* New subscribe command; make new Redis context
-				 * for this client */
-				cmd->ac = pool_connect(c->w->pool, cmd->database, 0);
-				c->pub_sub = cmd;
-				cmd->pub_sub_client = c;
+			if(ws->cmd != NULL) {
+				/* This client already has its own connection to Redis
+				   from a previous command; use it from now on. */
+
+				/* free args for the previous cmd */
+				cmd_free_argv(ws->cmd);
+				/* copy args from what we just parsed to the persistent command */
+				ws->cmd->count = cmd->count;
+				ws->cmd->argv = cmd->argv;
+				ws->cmd->argv_len = cmd->argv_len;
+				ws->cmd->pub_sub_client = c; /* mark as persistent, otherwise the Redis context will be freed */
+
+				cmd->argv = NULL;
+				cmd->argv_len = NULL;
+				cmd->count = 0;
+				cmd_free(cmd);
+
+				cmd = ws->cmd; /* replace pointer since we're about to pass it to cmd_send */
 			} else {
-				/* get Redis connection from pool */
-				cmd->ac = (redisAsyncContext*)pool_get_context(c->w->pool);
+				/* copy client info into cmd. */
+				cmd_setup(cmd, c);
+
+				/* First WS command; use Redis context from WS client. */
+				cmd->ac = ws->ac;
```

Comment:
here is where `cmd->ac` is set, so I guess it won't leak in `cmd_run`, but it also wouldn't use the correct one. And then one of the two would leak?

Reply:
See response above.